dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs

```diff
@@ -0,0 +1,254 @@
+        string question,
+        CancellationToken cancellationToken = default)
+    {
+        string planResponse = string.Empty;
+
+        if (string.IsNullOrEmpty(question))
+        {
+            throw new SKException("Goal not found.");
+        }
+
+        // Request completion for initial plan
+        var chatHistoryForPlan = await this.BuildChatHistoryForInitialPlanAsync(question, cancellationToken).ConfigureAwait(false);
+        string initialPlan = (await this._chatCompletion.GenerateMessageAsync(chatHistoryForPlan, null /* request settings */, cancellationToken).ConfigureAwait(false));
+
+        var chatHistoryForSteps = await this.BuildChatHistoryForStepAsync(question, initialPlan, cancellationToken).ConfigureAwait(false);
+        string resultContent = string.Empty;
+
+        for (int i = 0; i < this.Config.MaxIterations; i++)
+        {
+            // sleep for a bit to avoid rate limiting
+            if (i > 0)
+            {
+                await Task.Delay(this.Config.MinIterationTimeMs, cancellationToken).ConfigureAwait(false);
+            }
+
+            // For each step, request another completion to select a function for that step
+            chatHistoryForSteps.AddUserMessage(StepwiseUserMessage);
+            var chatResult = await this.GetCompletionWithFunctionsAsync(chatHistoryForSteps, cancellationToken).ConfigureAwait(false);
+            chatHistoryForSteps.AddAssistantMessage(chatResult);
+
+            OpenAIFunctionResponse? functionResponse = chatResult.GetOpenAIFunctionResponse();
+            if (functionResponse is null)
+            {
+                // The model replied with a message - check for final answer
+                var messageContent = (await chatResult.GetChatMessageAsync(cancellationToken).ConfigureAwait(false)).Content;
+
+                Match finalAnswerMatch = s_finalAnswerRegex.Match(messageContent);
+                if (finalAnswerMatch.Success)
+                {
+                    return new FunctionCallingStepwisePlannerResult
+                    {
+                        FinalAnswer = finalAnswerMatch.Groups[1].Value.Trim(),
+                        ChatHistory = chatHistoryForSteps,
+                        Iterations = i + 1,
+                    };
+                }
+
+                // No final answer found yet, so continue looping through steps
+                continue;
+            }
+
+            // Look up SKFunction
+            if (!this._kernel.Functions.TryGetFunctionAndContext(functionResponse, out ISKFunction? pluginFunction, out ContextVariables? funcContext))
+            {
+                throw new SKException($"Function {functionResponse.PluginName}.{functionResponse.FunctionName} not found in kernel.");
+            }
+
+            // Execute function and add to chat history
+            var result = (await this._kernel.RunAsync(funcContext, cancellationToken, pluginFunction).ConfigureAwait(false)).GetValue<object>();
```

Comment:
Did you consider catching this if execution fails and adding that as a user message? I'm not positive what the OpenAI guidance might be here.

Reply:
I don't believe OpenAI provides guidance on what to do in this case (as it's on the app to handle execution)... but this is a great idea! I'm already seeing improvements in success rate by giving the planner a chance to try something different in case of failure. Thanks!